packages/grid/x-data-grid-premium/src/concept/DataGrid.tsx

```diff
@@ -0,0 +1,606 @@
+    const firstRowRequest = this.screenRange.minY - this.cellsAround.up
+    const lastRowRequest  = this.screenRange.maxY + this.cellsAround.down
+
+    const firstRow = Math.max(0, firstRowRequest - (firstRowRequest % BLOCK_ROW_SIZE))
+    const lastRow = Math.min(this.rows.length, lastRowRequest + (BLOCK_ROW_SIZE - lastRowRequest % BLOCK_ROW_SIZE))
+
+    const firstColumnRequest = this.screenRange.minX - this.cellsAround.right
+    const lastColumnRequest  = this.screenRange.maxX + this.cellsAround.left
+
+    const firstColumn = Math.max(0, firstColumnRequest - (firstColumnRequest % BLOCK_COL_SIZE))
+    const lastColumn  = Math.min(this.columns.length, lastColumnRequest + (BLOCK_COL_SIZE - lastColumnRequest % BLOCK_COL_SIZE))
+
+    this.displayRange = {
+      minX: firstColumn,
+      maxX: lastColumn,
+      minY: firstRow,
+      maxY: lastRow,
+    }
+  }
+
+  updateDimensions() {
+    const scrollVerticalContent = this.scrollVertical.current!.children[0] as HTMLElement;
+    scrollVerticalContent.style.height = `${this.rowHeight * this.rows.length}px`
+
+    const rowHeight  = this.rowHeight;
+    const totalWidth = this.columns.length * this.columnWidth;
+
+    const borderColor = 'rgba(255, 255, 255, 0.3)';
+    const svg = `
+      <svg width="${totalWidth}" height="${rowHeight}" viewBox="0 0 ${totalWidth} ${rowHeight}" xmlns="http://www.w3.org/2000/svg">
+        <line x1="0" y1="${rowHeight}" x2="${totalWidth}" y2="${rowHeight}" stroke="${borderColor}" />
+      </svg>
+    `
+    const blob = new Blob([svg], { type: 'image/svg+xml' });
+    const url = URL.createObjectURL(blob); // XXX: 
+
+    const content = this.content.current!;
+    content.style.height = `${this.rowHeight * this.rows.length}px`;
+    content.style.width = `${totalWidth}px`;
+    content.style.backgroundImage = `url(${url})`
+    content.style.backgroundSize = `${totalWidth}px ${rowHeight}px`
+    content.style.backgroundRepeat = 'repeat-y';
+
+    if (this.backgroundImageURL) {
+      URL.revokeObjectURL(this.backgroundImageURL);
+    }
+    this.backgroundImageURL = url
+  }
+
+  renderBlock(x: number, y: number) {
+    const rowIndex = y
+    const columnIndex = x
+
+    const element = document.createElement('div')
+    element.className = 'block'
+    element.setAttribute('data-row', String(rowIndex))
+    element.setAttribute('data-col', String(columnIndex))
+
+    const xPx = columnIndex * this.columnWidth
+    const yPx = rowIndex * this.rowHeight
+    element.style.transform = `translate3d(${xPx}px, ${yPx}px, 0)`
+    // element.style.top = `${rowIndex * this.rowHeight}px`
+    // element.style.left = `${columnIndex * this.columnWidth}px`
+    // element.style.containIntrinsicWidth = `${BLOCK_COL_SIZE * this.columnWidth}px`
+    // element.style.containIntrinsicHeight = `${BLOCK_ROW_SIZE * this.rowHeight}px`
+
+    this.content.current!.appendChild(element)
+
+    const root = ReactDOM.createRoot(element)
+    root.render(
```

Comment:
Interesting, I think this is the first time I see multiple React roots being used like this.
I wonder how would we update `<Block/>` components?

I created this demo to see if I can share context between multiple React roots, and it didn't work:
https://codesandbox.io/s/epic-framework-y9cmmx?file=/src/App.tsx
Am I missing something here?

Reply:
We don't need to update blocks with this implementation, they need mounting & unmounting only. The reactive updates of the grid would be handled by our existing mechanism with `useGridSelector`, which can work even across roots.

However, it's true that custom user context doesn't work across roots.

An alternative here could be to use portals instead of roots.

Reply:
> However, it's true that custom user context doesn't work across roots.

I guess roots are a no-go then - they need to have access to the theme context among others:
https://codesandbox.io/s/eager-jennings-yxcjcz?file=/src/App.tsx

Reply:
> An alternative here could be to use portals instead of roots.

Is it possible to imperatively mount/unmount portals though?

Reply:
> I guess roots are a no-go then - they need to have access to the theme context among others:

Depends if we accept to support only some contexts. It's easy to manually re-wrap the subroots with more contexts such as the theme one, but there is no general solution to support all contexts afaik.

> Is it possible to imperatively mount/unmount portals though?

Not sure, probably not :|

Reply:
> It's easy to manually re-wrap the subroots with more contexts such as the theme one, but there is no general solution to support all contexts afaik.

Yeah, that would work for the MUI theme context.
But it won't be possible to use other component libraries that rely on context in cell renderer ([Material UI could be used with other component libraries](https://mui.com/material-ui/guides/styled-engine/#theme-scoping))